Add tests for the existing config

williammartin opened the pull request:
### Description

In preparation for changing our configuration schema to support multiple accounts, I want gain confidence in our Config, particularly the pieces that relate to the `AuthConfig` structure and reach down into the config and keyring. I actually couldn't find any tests that provided coverage for this even as an integration.

There will need to also be some tests added to the `go-gh/auth` package which is called through here and reaches into the `config`.

I've left some comments on areas I want to call out.

williammartin committed:
Add initial tests for config Get

williammartin committed:
Add tests for config GetOrDefault

williammartin committed:
Add initial AuthConfig tests

williammartin committed:
Add tests for AuthConfig Login and Logout

williammartin committed:
Update AuthConfig to use default fn

williammartin committed:
Add tests for AuthConfig Token

williammartin committed:
Mod tidy go-md2man

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,302 @@
+
+	// When the user has logged in insecurely
+	authCfg := newTestAuthConfig()
+	ghConfig.Read = func() (*ghConfig.Config, error) {
```
A bit sad 😬

samcoe replied:
I dont think this is necessary:
```go
	cfg, err := NewConfig()
	require.NoError(t, err)
	authCfg := cfg.Authentication()
```

williammartin replied:
I don't think what you're proposing here actually resolves my sadness so just to be clear about why I'm doing this `NewConfig` calls to `ghConfig.Read` which uses `sync.Once` to load config from disk. That means that any test from that point onward that calls `ghConfig.Read` will not be isolated.

samcoe replied:
Ah I see. I misunderstood the purpose. That makes sense to me.

williammartin replied:
I think what I'll do is actually put it inside the `newTestAuthConfig` function or another new function and comment it so that the reason can be discovered from any test.

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,302 @@
+	require.ErrorIs(t, err, keyring.ErrNotFound)
+}
+
+// Note that I'm not sure this test enforces particularly desirable behaviour
```
Calling this out for attention.

samcoe replied:
The reasoning behind this behavior was that if logout fails what actions can the user actually take? The thought was that there really was none and thus we shouldn't show them an error. Not saying I agree, just some context.

williammartin replied:
Depends on the user I suppose. A user that cares that their token is still hanging around on a shared machine could care.

I don't feel a pressing need to change it but I wanted to leave a note for any maintainer that comes across this in future. In fact, I'll update the comment to capture your added context.

williammartin commented on internal/config/config.go:
```diff
@@ -53,22 +53,13 @@ func (c *cfg) Get(hostname, key string) (string, error) {
 		return val, err
 	}
 
-	if defaultExists(key) {
```
I dunno, feel free to tell me to move it back but it felt pretty analogous to a map.

samcoe replied:
I like the change.

williammartin commented on internal/config/config.go:
```diff
@@ -178,13 +160,19 @@ func (c *AuthConfig) User(hostname string) (string, error) {
 
 // GitProtocol will retrieve the git protocol for the logged in user at the given hostname.
 // If none is set it will return the default value.
+// TODO: although this returns an error, it actually has no path to error.
```
Calling this out for attention.

samcoe replied:
Let's change the function signature then, it is only used in two places. Interestingly enough this function should replace the majority of places `GetOrDefault` is used. Would love to clean that up as I find `GetOrDefault` to be a bit of an eyesore.

williammartin commented on internal/config/config.go:
```diff
@@ -245,6 +235,8 @@ func (c *AuthConfig) Login(hostname, username, token, gitProtocol string, secure
 // Logout will remove user, git protocol, and auth token for the given hostname.
 // It will remove the auth token from the encrypted storage if it exists there.
 func (c *AuthConfig) Logout(hostname string) error {
+	// TODO: I can't find anywhere that expects to call this function without a hostname
```
Calling this out for attention.

samcoe replied:
I think this was in place for legacy reasons, probably okay to remove now.

williammartin commented on internal/config/config.go:
```diff
@@ -233,6 +221,8 @@ func (c *AuthConfig) Login(hostname, username, token, gitProtocol string, secure
 		c.cfg.Set([]string{hosts, hostname, oauthToken}, token)
 		insecureStorageUsed = true
 	}
+	// TODO: I can't find anywhere that calls this function without a username, so that would seem like
```
Calling this out for attention.

samcoe replied:
I think this was in place for legacy reasons, probably okay to remove now.

williammartin committed:
Add tests for AuthConfig HasEnvToken

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,302 @@
+package config
```
Note that I have not written tests for `KnownHosts` (because it has no extra behaviour on top of `go-gh/auth` right now so it was low priority) nor the `AliasConfig` because it is low risk for the multi account changes.

williammartin replied:
I have written tests for `Hosts` and `DefaultHost` now but I will leave `AliasConfig` out of this PR and do it in a follow up since this is already getting pretty hefty.

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,350 @@
+}
+
+func TestHasEnvTokenWithNoEnvTokenButAConfigVar(t *testing.T) {
+	t.Skip("this test is explicitly breaking some implementation assumptions")
```
Calling this out for attention.

williammartin commented on internal/config/config.go:
```diff
@@ -153,6 +135,14 @@ func (c *AuthConfig) HasEnvToken() bool {
 			return true
 		}
 	}
+	// TODO: This is _extremely_ knowledgable about the implementation of TokenFromEnvOrConfig
```
Calling this out for attention.

samcoe replied:
Agreed. We should create a function `HasEnvToken` in `go-gh` that abstracts away this knowledge.

williammartin replied:
I'm going to punt that from this PR since it's a bigger change and this is already pretty hefty.

williammartin commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,350 @@
+	// Then it returns success, and stores the user and git protocol in the config
+	require.NoError(t, err)
+
+	requireKeyWithValue(t, authCfg.cfg, []string{hosts, "github.com", "user"}, "test-user")
+	requireKeyWithValue(t, authCfg.cfg, []string{hosts, "github.com", "git_protocol"}, "ssh")
```
I should test these using `GitProtocol` and `User` instead. That would be a better black box test.

samcoe commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,350 @@
+
+	// When the user is authenticated via env var
+	authCfg := newTestAuthConfig()
+	ghConfig.Read = func() (*ghConfig.Config, error) {
```
Same comment as above.

samcoe commented on internal/config/config.go:
```diff
@@ -53,22 +53,13 @@ func (c *cfg) Get(hostname, key string) (string, error) {
-	}
-
-	val, err = c.cfg.Get([]string{key})
+	val, err := c.Get(hostname, key)
```
Nice 👍

samcoe commented on internal/config/config.go:
```diff
@@ -153,6 +135,14 @@ func (c *AuthConfig) HasEnvToken() bool {
+	// can guarantee that tokens will only be returned from a set env var.
+	// Discussed here, but maybe worth revisiting: https://github.com/cli/cli/pull/7169#discussion_r1136979033
+	//
+	// By providing example.com. it's also _only_ looking for GH_ENTERPRISE_TOKEN or GITHUB_ENTERPRISE_TOKEN
```
It checks `GH_TOKEN` and `GITHUB_TOKEN` as well.

williammartin replied:
Ohhh sorry, I missed that the `isEnterprise` branch of `tokenForHost` would fall through. Thanks for correcting me!

samcoe reviewed (commented):
@williammartin This is excellent, thanks for being diligent about improving our test coverage! I left inline code comments, let me know if you have any questions about them.

williammartin committed:
Rewrite AuthConfig User and GitProtocol tests

williammartin committed:
Add additional comment on Logout ignoring errors

williammartin committed:
Add test that host is added on Login

williammartin committed:
Share config.Read setup in AuthConfig tests

williammartin committed:
Add tests for AuthConfig DefaultHost

williammartin committed:
Avoid using keyring unnecessarily in AuthConfig tests

This is to avoid the impression that the keyring matters, when the test
is actually about general auth behaviour

williammartin committed:
Ensure all AuthConfig tests are isolated on disk

williammartin committed:
Remove Logout branch that handles empty hostname

williammartin committed:
Remove Login branch that handles empty username

williammartin committed:
Remove incorrect comment about HasEnvToken

williammartin committed:
Fix GitProtocol signature and rework test descriptions

williammartin committed:
Remove GetOrDefault uses in favour of GitProtocol

williammartin commented:
I believe I've addressed all comments @samcoe and have explicitly excluded:
 * Testing `AliasConfig`
 * Moving `HasEnvToken` into `gh-auth`

In both cases, I felt like this PR was already pretty hefty and they were both larger distractions from the multi account confidence that I wanted to get out of this.

samcoe commented on internal/config/auth_config_test.go:
```diff
@@ -0,0 +1,377 @@
+
+	// Then the token is successfully fetched
+	// and the source is set to oauth_token but this isn't great
+	// but I can't find the issue # that references this.
```
I think https://github.com/cli/go-gh/issues/94 is what you are looking for.

williammartin replied:
Yes! I just could not find it anywhere

samcoe reviewed (approved):
Wonderful! Thanks for making the requested changes. This looks ready to ship. Glad we are one step closer to removing `GetOrDefault`.

williammartin committed:
Add link to token source issue
